### PR TITLE
Bucket online offline 1

### LINKED ISSUE #78: [DB Online / Offline] Scenario #1
Start SG with bucket in online state (no offline config property), make REST API calls against DB - expected result RAST API calls are successful

### LINKED ISSUE #79: [DB Online / Offline] Scenario #2
Start SG with bucket in online state setting Db 'offline' config property to false),  make REST API calls against DB - expected result REST API calls are successful

### LINKED ISSUE #80: [DB Online / Offline] Scenario #3
With DB in online state, no active REST API calls, take DB offline, expected result _offline call returns status 200, make REST API calls against DB, expected result returns status 503.

### LINKED ISSUE #81: [DB Online / Offline] Scenario #4
With DB in online state, make REST API call to get DB runtime details /db/, expected result 'state' property with value 'Online' is returned.

### LINKED ISSUE #83: [DB Online / Offline] Scenario #6
Start SG with bucket in offline state setting DB 'offline' config property to true),  make REST API calls against DB - expected result RAST API calls fail with 503

### LINKED ISSUE #93: [DB Online / Offline] Config Scenario #3
With DB in online state, put new invalid DB config, take DB offline, bring DB online. Expected result DB fails to come online, error status is returned.

### PR DESCRIPTION
**Pull request by sethrosetter**
- fixes #78 
- fixes #79 
- fixes #80 
- fixes #81 
- fixes #83 
- fixes #93 

- adds admin functionality
```
admin.get_roles(db)
admin.get_role(db, name)
admin.get_users_info(db)
admin.get_user_info(db, name)
admin.db_online(db)
admin.db_offline(db)
admin.get_db_config(db)
admin.put_db_config(db, config)
```
- adds user functionality
```
user.get_all_docs(),
user.start_polling(termination_doc_id, timeout)
user.start_continuous(termination_doc_id, timeout)
```

- add fixture to disable http retry.
- add REST scanner to verify db endpoints in online vs offline state
- add method to log.info for user.py and admin.py
- add sanity longpolling changes feed test
- add sanity continuous changes feed test

### CODE REVIEW

File: conf/bucket_online_offline/bucket_online_offline_offline_true.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    "slowServerCallWarningThreshold": 500,
+    "compressResponses": true,
+    "log":["*"],
+    "verbose":"true",
+    "databases":{
+        "db":{
+            "offline":true,
+            "server":"http://{{ couchbase_server_primary_node }}:8091",
+            "bucket":"data-bucket",
+            "users": { "seth": { "password": "password", "admin_channels": ["*", "ABC"] } }
```
Comment (ashvindersingh) [inline conf/bucket_online_offline/bucket_online_offline_offline_true.json]: Do we need the 'users' here for bucket-online-offline config?

**Comment by sethrosetter** [comment]
@ashvindersingh Lets get your web hook framework enhancements into this branch before we merge this to master unless you are branched off of master for that work.

File: lib/admin.py [hunk 2]
```diff
@@ -52,3 +77,54 @@ def register_bulk_users(self, target, db, name_prefix, number, password, channel
...
+        log.info("GET {}".format(resp.url))
+        resp.raise_for_status()
+        return resp.json()
+
+    # POST /{db}/_resync
+    def db_resync(self, db):
+        pass
+
+    # POST /{db}/_online
+    def db_online(self, db, delay=None):
```
Comment (ashvindersingh) [inline lib/admin.py]: rename to bring_db_online

File: lib/admin.py [hunk 3]
```diff
@@ -52,3 +77,54 @@ def register_bulk_users(self, target, db, name_prefix, number, password, channel
...
+        if delay is not None:
+            data = {"delay": delay}
+
+        resp = requests.post("{0}/{1}/_online".format(self.admin_url, db), headers=self._headers, timeout=settings.HTTP_REQ_TIMEOUT, data=json.dumps(data))
+        log.info("POST {}".format(resp.url))
+        resp.raise_for_status()
+        return resp.status_code
+
+    # POST /{db}/_offline
+    def db_offline(self, db):
```
Comment (ashvindersingh) [inline lib/admin.py]: rename to take_db_offline.

File: lib/user.py [hunk 4]
```diff
@@ -273,3 +294,96 @@ def get_changes(self, feed=None, limit=None, heartbeat=None, style=None,
...
+                obj = r.json()
+
+                new_docs = obj["results"]
+
+                # Check for duplicates in response doc_ids
+                doc_ids = [doc["doc"]["_id"] for doc in new_docs if not doc["id"].startswith("_user/")]
+                assert(len(doc_ids) == len(set(doc_ids)))
+
+                if len(new_docs) == 0:
+                    request_timed_out = True
```
Comment (ashvindersingh) [inline lib/user.py]: request_timed_out is not reflective that changes feed return zero docs.

**Pull request merged**